Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+    {
+    }
+
+    Value::Value(bool value)
+        : m_value(value)
+    {
+    }
+
+    Value::Value(Type type)
+    {
+        switch (type)
+        {
+        case Type::Null:
+            // Null is the default initialized value
+            break;
+        case Type::Bool:
+            m_value = false;
+            break;
+        case Type::Object:
+            SetObject();
+            break;
+        case Type::Array:
+            SetArray();
+            break;
+        case Type::String:
+            SetString("");
+            break;
+        case Type::Int64:
+            m_value = int64_t{};
+            break;
+        case Type::Uint64:
+            m_value = uint64_t{};
+            break;
+        case Type::Double:
+            m_value = double{};
+            break;
+        case Type::Node:
+            SetNode("");
+            break;
+        case Type::Opaque:
+            AZ_Assert(false, "AZ::Dom::Value may not be constructed with an empty opaque type");
+            break;
+        }
+    }
+
+    Value& Value::operator=(const Value& other)
+    {
+        m_value = other.m_value;
+        return *this;
+    }
+
+    Value& Value::operator=(Value&& other) noexcept
+    {
+        m_value.swap(other.m_value);
+        return *this;
+    }
+
+    bool Value::operator==(const Value& rhs) const
+    {
+        if (IsString() && rhs.IsString())
+        {
+            return GetString() == rhs.GetString();
+        }
+        else if (IsNumber() && rhs.IsNumber())
+        {
+            if (IsInt())
+            {
+                return GetInt64() == rhs.GetInt64();
+            }
+            else if (IsUint())
+            {
+                return GetUint64() == rhs.GetUint64();
+            }
+            else
+            {
+                return GetDouble() == rhs.GetDouble();
+            }
+        }
+        else
+        {
+            return m_value == rhs.m_value;
+        }
+    }
+
+    bool Value::operator!=(const Value& rhs) const
+    {
+        return !operator==(rhs);
+    }
+
+    void Value::Swap(Value& other) noexcept
+    {
+        m_value.swap(other.m_value);
+    }
+
+    Type Dom::Value::GetType() const
+    {
+        switch (m_value.index())
+        {
+        case 0: // AZStd::monostate
+            return Type::Null;
+        case 1: // int64_t
+            return Type::Int64;
+        case 2: // uint64_t
+            return Type::Uint64;
+        case 3: // double
+            return Type::Double;
+        case 4: // bool
+            return Type::Bool;
+        case 5: // AZStd::string_view
+        case 6: // AZStd::shared_ptr<const AZStd::string>
+        case 7: // ShortStringType
+            return Type::String;
+        case 8: // ObjectPtr
+            return Type::Object;
+        case 9: // ArrayPtr
+            return Type::Array;
+        case 10: // NodePtr
+            return Type::Node;
+        case 11: // AZStd::any*
+            return Type::Opaque;
```

Comment:
Instead of typecasting this to magic integer numbers, why not just use [AZStd::holds_alternative](https://en.cppreference.com/w/cpp/utility/variant/holds_alternative) and compare against the actual types.

i.e
```
if (AZStd::holds_alternative<AZStd::monostate>(m_value))
{
    return Type::Null;
}
else if (AZStd::holds_alternative<...>(m_value))
{
    return Type::...;
}
```

Reply:
Tried this, saw a slight perf regression (which was unintuitive, I'd have thought it would optimize per @AMZN-koppersr 's comment about optimizing to a jump table), didn't feel like digging on that so I wrote a `GetTypeIndex` helper to translate a type to an index in a safe manner for this. There's probably some STD helper for extracting template parameters I'm missing that would make this more idiomatic, implementation suggestions welcome.

Reply:
There is [AZStd::get](https://en.cppreference.com/w/cpp/utility/variant/get) and [AZStd::get_if](https://en.cppreference.com/w/cpp/utility/variant/get_if) which can be used to extract an alternative from a variant.